File: 189-rotate-array/rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n=nums.size();
       if(k%n==0) {
           return;
       }
       //if k is greater than n than we only have to the remainder when size is divided. becoz rotating the array n time gives the same array
       if(k>n){
           k=k%n;
       }
        reverse(nums.begin(),nums.end()-k);
        reverse(nums.end()-k,nums.end());

        reverse(nums.begin(),nums.end());
    }
};

File: 189-rotate-array/rotate-array.java
class Solution {
    public void rotate(int[] arr, int k) {
        int size = arr.length;
        k = k % size;
        reverseArray(arr,0,size - k-1);
        reverseArray(arr,size-k,size-1);
        reverseArray(arr,0,size-1);
    }

    private void reverseArray(int[] arr, int start, int end){
        while(start < end){
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end]=temp;
            start++;
            end--;
        }
    }
}